### PR TITLE
Replace VidScheme assoc type aliases with newtype wrappers

### LINKED ISSUE #3790: [TECH_DEBT] - VidSchemeType associated types should also be opaque
### What is this task and why do we need to work on it?

Currently we hide the implementation details of the VID scheme in `crates/types/src/vid.rs`. As documented in that module, we expose only a constructor function `vid_scheme` that returns a newtype wrapper `VidSchemeType` that impls jellyfish traits such as `VidScheme`. (We need a newtype because Rust forbids naming impl trait in return types. ☹ ) So far so good.

Unfortunately, we still expose the underlying type for all the associated types for the `VidScheme` trait: `Commit`, `Common`, `Share`:
https://github.com/EspressoSystems/HotShot/blob/53e28ead5eebf1bc4e713ec1a5d6f580d33db2cf/crates/types/src/vid.rs#L200-L203

Sure enough, we already have code that relies on the fact that our `VidScheme::Commit` type is a `GenericArray`. 🤦 Specifically, the jellyfish `VidScheme` trait does not specify any `AsRef` trait bound for the `Commit` assoc type, yet we rely on it in several places. Examples:
- in production code: https://github.com/EspressoSystems/HotShot/blob/53e28ead5eebf1bc4e713ec1a5d6f580d33db2cf/crates/types/src/data.rs#L322
- in production code: we assume `VidScheme::Commit` is `Copy` pretty much everywhere.
- in test code: https://github.com/EspressoSystems/HotShot/blob/53e28ead5eebf1bc4e713ec1a5d6f580d33db2cf/crates/example-types/src/block_types.rs#L398-L400

(It sure would be nice if Rust ergonomics would make abstraction the default, but alas...)

I'm now running into trouble in my efforts to fix #3773 : I want to swap out ADVZ for a dummy VID scheme. I find myself jumping through hoops to make `.as_ref().as_ref()` work in my dummy VID scheme.

Better to solve this problem properly now instead of piling on more cruft.

### What work will need to be done to complete this task?

Make new newtypes for `VidScheme` assoc types `Commit`, `Common`, `Share` that expose only the traits declared for these types in jellyfish. Enforce hotshot talk to these types only through their newtype wrappers, just like we currently do with `VidSchemeType`.

### Are there any other details to include?

_No response_

### What are the acceptance criteria to close this issue?

Just review.

### Branch work will be merged to (if not the default branch)

_No response_

### PR DESCRIPTION
**Pull request by ggutoski**
Closes #3790 

### This PR: 

- As described in the issue, make new newtypes for `VidScheme` assoc types `Commit`, `Common`, `Share`, `PrecomputeData` that expose only the traits declared for these types in jellyfish. Enforce hotshot talk to these types only through their newtype wrappers, just like we currently do with `VidSchemeType`.
- All core changes confined to `crates/types/src/vid.rs`. All other changes accommodate the changes in this file.
- `VidCommitment` was especially bad. Everywhere we assume this type is `AsRef` and `Copy`. I replaced each use of `AsRef` with `bincode::serialize` and each copy with `.clone()`. All other `VidScheme` assoc types were trivial to change.

### Key places to review: 

- IMPORTANT: Is the replacement of `as_ref()` with `bincode::serialize` backward compatible? ie. Do both these calls yield the same bytes? I think so but don't know for sure.
- Frequent use of `bincode::serialize` is ugly. `bincode` is only way upstream (jellyfish) allows to convert a VID commitment into a `[u8]`. Not sure what the best solution is here. I'm open to suggestion.
- Add `derive_more` dependency to `types` crate. The workspace-level dependency lacks feature `display`, so I added it to `crates/types/Cargo.toml`. Perhaps we should simply add the `display` feature to the workspace-level dependency?

### CODE REVIEW

File: crates/types/src/traits/signature_key.rs [hunk 1]
```diff
@@ -336,7 +336,10 @@ fn aggregate_fee_data<Metadata: EncodeBytes>(
     let mut fee_info = Vec::new();
     fee_info.extend_from_slice(fee_amount.to_be_bytes().as_ref());
     fee_info.extend_from_slice(metadata.encode().as_ref());
-    fee_info.extend_from_slice(vid_commitment.as_ref());
+    fee_info.extend_from_slice(
+        &bincode::serialize(&vid_commitment)
```
[the remark this answers is not in the document]
Comment (ggutoski) [reply crates/types/src/traits/signature_key.rs:340]: It is possible. But `AsRef` is not something jellyfish promises in trait bounds. Hotshot code had to break abstraction in order to use it. (ie. Hotshot code relied on underlying implementation details of ADVZ commitment to access `as_ref`.)

Supporting `AsRef` is troublesome because it returns a *reference*, which basically forces the underlying implementation to store its data as a byte slice, as anything else would require a conversion to `[u8]` that immediately goes out of scope.

`AsRef` isn't really meant for serialization, which is what we actually want here. Instead, `AsRef` is intended only for "[cheap reference-to-reference conversion](https://doc.rust-lang.org/std/convert/trait.AsRef.html)".

## Options
1. Bite the bullet and impose a `AsRef` requirement. If we do this then ultimately we should do it upstream, ie. add `AsRef` to the trait bounds promised by jellyfish. If we choose this option then I'm willing to restore support for `AsRef` in this PR in anticipation of a future change to jellyfish API.
2. Add a helper function to hotshot that serializes a `VidCommitment` into a `Vec<u8>`. This removes the need for the ugly `bincode::serialize(...).expect(...)` everywhere. It's still there, of course. It just appears in one place instead of everywhere. Naturally, this option is slower because bincode needs to allocate heap memory and copy into it.

[the remark this answers is not in the document]
Comment (ss-es) [reply crates/types/src/traits/signature_key.rs:340]: Oh, I think I understand the ownership issue with `AsRef`.

I guess I'm less concerned about the cost of the copy (it's only a few hundred bytes I think?) as I am about the `bincode` and the `expect`. Is it possible for jellyfish to add an extra trait bound that exports an infallible `VidCommitment -> Vec<u8>`?

[the remark this answers is not in the document]
Comment (ggutoski) [reply crates/types/src/traits/signature_key.rs:340]: Well, a lot of `Vec` operations could panic, too. I don't see any reason why panic via bincode is more likely than panic via Vec. bincode is fallible because it's extremely general and serde-compatible. In our application I expect it never to fail, so panic-on-failure is probably an acceptable solution. I confess I don't like the panic either, but that seems to be the nature of Rust.

If we were to add an infallible conversion to `Vec<u8>` then we'd be duplicating the functionality of serde. (Indeed, I would probably implement this "infallible" conversion via `bincode` + `expect` just to ensure compatibility with serde. 😛 )

Bottom line: if your only objection to bincode is that it might panic then I wouldn't worry about it. TBH my main objection is performance. Seems like we can't avoid touching the heap and copying bytes each time we access the commitment. It hurts my OCD not to break abstraction and return a reference to internal bytes. 😬

[the remark this answers is not in the document]
Comment (ggutoski) [reply crates/types/src/traits/signature_key.rs:340]: My suggestion: add to `types/src/vid.rs` a `pub fn` to wrap the ugly `bincode` + `expect`. We could call it `as_bytes` or even `bincode_serialize`.

[the remark this answers is not in the document]
Comment (ggutoski) [reply crates/types/src/traits/signature_key.rs]: On further consideration, we decided to restore use of `AsRef`. On balance, the benefits (ergonomics, efficiency) outweigh the cost of lost abstraction. Done in https://github.com/EspressoSystems/HotShot/pull/3794/commits/d74cc908174ab3ff2a824d5d29111ca9098620c4

File: crates/example-types/src/block_types.rs [hunk 2]
```diff
@@ -394,10 +394,8 @@ impl Committable for TestBlockHeader {
                 <TestBlockHeader as BlockHeader<TestTypes>>::block_number(self),
             )
             .constant_str("payload commitment")
-            .fixed_size_bytes(
-                <TestBlockHeader as BlockHeader<TestTypes>>::payload_commitment(self)
-                    .as_ref()
-                    .as_ref(),
+            .var_size_bytes(
+                <TestBlockHeader as BlockHeader<TestTypes>>::payload_commitment(self).as_ref(),
```
Comment (ss-es) [inline crates/example-types/src/block_types.rs:398]: small question: was this change `fixed_size_bytes` -> `var_size_bytes` forced by the refactor?

it doesn't really matter here (`example-types` is completely internal), I just wanted to check whether the sequencer is going to be forced to change the commitment. or does the commitment work out to the same thing anyway?

**Review by ss-es — APPROVED** [review]
looks good to me, just one question

File: crates/example-types/src/block_types.rs (hunk 2 above)
Comment (ggutoski) [reply crates/example-types/src/block_types.rs:398]: > small question: was this change `fixed_size_bytes` -> `var_size_bytes` forced by the refactor?

Yes. Previously this code exploited `AsRef<[u8; 32]>` whereas now we support only `AsRef<[u8]>`; hence the switch.

At first I worried that this could break backward compatibility. But as you observe it affects only example-types in this repo so I decided it's fine.

I did not consider the sequencer though. I guess we should check this. Let's discuss offline.

Comment (ggutoski) [reply crates/example-types/src/block_types.rs:398]: This change does indeed break backward compatibility of downstream code. 😢 As discussed offline, the path of least resistance is to support `AsRef<[u8; 32]>` instead of `AsRef<[u8]>`. Done in https://github.com/EspressoSystems/HotShot/pull/3794/commits/37ec30ca80cad1a33c985cfcd28685934b0b2ba2.
